README.md
# Cockos Happy AVS
This is an annotated version of [Cockos Happy AVS](http://visbot.github.io/AVS-Forums/html/t-321482.html).

## Setup
This is a setup that can build the code. It could probably work with different versions.
+ Visual C++ 6.0
+ Windows XP SP3
+ Winamp 5.58

## Build
1. Open workspace vis_avs/vis_Avs.dsw
2. Set Active build configuration to Win32 Release
3. Build vis_chavs.dll (F7). The link settings outputs the dll directly to the winamp plugins directory. Modify in project settings if required.

## Debug
1. Open workspace vis_avs/vis_Avs.dsw
2. Set Active build configuration to Win32 Debug
3. Set winamp executable in project settings
2. Start Debug (F5)

## Nullsoft Advanced Visualization Studio (AVS) v2.0a4 README

Thanks to Paul Holden for the FunkyFX Firewurx APE  

Once you install this baby, fire up Winamp, open up the preferences 
(CTRL+P), go to the Plugins/Visualization section, and select 
Nullsoft AVS as the visualization plug-in. Hit start. 

Once the AVS opens, it will probably be showing a nice black output.
To get AVS to display something more interesting, hit space, or right
click in the black area and select a preset to load.

If you want to create your own presets, click the left mouse button in 
the black area of the window to bring up the AVS Editor window.

The editor lets you create new visuals by adding effects from your Effect
Library (in the upper right) to the Active Effect list (on the left side).
Once you've added effects, you can configure each effect by selecting them
from the Active Effect list. You can also clear the list, or load or save
the list to be loaded later as a preset. 

You can also configure some of AVS's settings by selecting items from the 
Settings section of the Editor. 

Hotkeys for main window:
 *  Any winamp key
 *  R toggles randomswitching
 *  F toggles fullscreen framerate counter
 *  Y and U cycle through presets in order
 *  Space goes to random preset
 *  Enter toggles fullscreen
 *  0,1-9, F1-F10 load presets
 *  Ctrl+above save presets

That's it for now, and enjoy!

HISTORY.txt
-francis:
added floor(), ceil() and $Phi
added "cancel fullscreen on deactivation" toggle option

-mig:
added example infrastructure to dmove.  :)
now there are 8 examples in dmove.  also tuned up my beat-interpolation algos!  woo!

-justin
added blend modes to misc / buffer save
made config expand trees by default

-mig:
started work on undo/redo system.
infrastructure seems to work, but I can't figure out the magical
  incantation to make the UI reflect the newly altered render list
  although it seems to actually go back to the last values in the renderer.
"Do you want to save?" dirty flag now contained (mostly) in the undo object.
(In fact, it saves a dirty bit for every undo state, so you can undo twice,
  save, then redo and you should have your dirty bit still set.  If you then
  undo yet again, you're back to what you saved and you have no dirty bit)
added more examples to SuperScope

-justin:
fixed memory leak in evallib stuff I made a bit back
added debug window, moved register watch to it, added error listing
added option for clear variables on recompile in debug window
fixed bugs, too.

-justin:
started adding setting for reinit on edit for evallib stuff
made config change detection a little more accurate
made superscope/dmove/shift/etc reinit on edit work right
made trans / movement only replace usereval with preseteval if editted
fyi: trans / movement also provides 'sw' and 'sh' for screen width and height,
  so things like 'd=d*0.93 + (((x*sw/2)&(y*sh/2)&1)*-0.3);' are fun :)
fixed trans / movement backwards compatible stuff (no crashy at least)
updated eval documentation stuff, need people to write docs for it, too.. heheh
lots of ui cleanups (more edit room!)
fixed some fullscreen new fbresize bugs
updated fullscreen overlay mode some
updated maximum local variables to 256 from 64 
added global registers [reg00,reg01,... reg99] for debug 
  and effect synchronization (or inter-preset communication) 
added global register watch in root list window (for debugging) 
added comments for eval code. // and /* bla */ 
updated eval documentation 
added gettime() function to eval 
made annoying grey rectangle below tree go away when you aren't docked 
added new color modifier effect, which can be used to replace 
  (with script) a lot of different effects (and be faster and more flexible). 
added 'minimum blend' mode for effect lists/line drawing 
added 'round up' option for Blur. 
made a generic avs directory scanner routine for subdirs (yay!)
   made it only generate menus when opened, for speed
made effects that use evallib reset variables to empty on recompile
made avs track dirtiness of presets, and (optionally) prompt to save etc
made avs resize images when resizing (optionally)
added extended APE interface, that lets you access evallib, global registers,
  line blend/width modes. (see apesdk.zip)




- mig: made most of the presets in Trans/Movement show their algos in the edit field (so you can tune them yourself)
       added multiple new algos to the preset list

- mig: removed ability to RMB in fullscreen mode.  Bug 331 from the database.  want make new gui.  want make now.
- basu: added 'onbeat enabled' option to lists
- christophe: added thread priority settings (so you can set avs in idle prority while keeping winamp in
  high prio.)
2.0a7:
- made better multimonitor support -- still needs more work (making it able to go fullscreen and you do other things
2.0a6:
- added 'skip first' in misc/custom bpm
- added interferences effect
2.0a4:
- auto fullscreen window resizing
2.0a3:
- updated fullscreen vis code to handle leaving fullscreen
  unexpectedly better
- made configwnd repopulation more robust
- made transitions work when preinit is disabled and you just
  started up
2.0a2:
- updated options of display/fullscreen/trans
- added doublesize for windowed
- added seperate opts for windowed/fullscreen for text
- fixed bug in mosaic.

2.0a1:
- preinit of presets for better transitions
- integrated laser support (built-time option)
- changed name to 'Winamp AVS'
- Improved SVP/UVS loading
- Made superscope support 'red' 'green' and 'blue'

1.5a6:
- preset transitions, woohoo
- yay

1.5a4:
- presets for superscope
- more superscope options
- made random preset loading on beat
- bugfixes

1.5a3:
- line width option
- ability to use buffer as alpha channel for blending sub-effects back in
- tons more

1.5a2:
- deadbeef: cleanups
- lone: adjustable blend for sub-effects

1.5a1:
- deadbeef: made new effect system completely hiererchical and scaleable. 
			reorganized code a lot. cleaned things up. made it nice.

1.0a53:
- deadbeef: superscope effect
			made it save to plugins\vis_avs.dat, instead.
-lone :		added Trans / Invert
			added Trans / Unique tone
			added Render / Timescope

a52:
- deadbeef: optimized/simplified evallib. made it limited to 8 char variable names,
			32 variables max.
			improved ddm effect.
			improved color clip effect

a51:
- deadbeef:	optimized mosaic, grain, brightness, and bump effects
			optimized and added more functionality to interleave effect
- lone :	clear - fixed 'first frame only'
			eval - added sigmoid, sign, max, min, rand, band, bor, bnot, if, equal, above, below
			ddm  - added code for init and beat
			bump - added 'bi' var to control bump intensity thru exps.
			     - added depth source
			clear - fixed 'first frame only' (again)
			onbeat clear - fixed 'skip n beats' which was not saved
- ron : picture - fixed picture border bugs when aspect ratio was on

a50:
- deadbeef: added subtractive blend, every other line blend to stack
			fixed window-no-erase bug.
			added new dynamic distance modifier effect
			added 'go' button to fullscreen options
			added wait for retrace options
			revised logarithmic spectrum scaling table
- ron: better no-minimize-on-winamp (now displays a separate window in taskman)
- lone : bpm - better out of range detection in average calculation
						 - better confidence calculation

vis_avs/render.cpp
/*
  LICENSE
  -------
Copyright 2005 Nullsoft, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, 
are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer. 

  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution. 

  * Neither the name of Nullsoft nor the names of its contributors may be used to 
    endorse or promote products derived from this software without specific prior written permission. 
 
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR 
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND 
FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

*/
#include <windows.h>
#include "render.h"
#include "timing.h"
#include "undo.h"

#ifdef LASER
C_LineListBase *g_laser_linelist;
#endif

C_RenderListClass *g_render_effects;
C_RenderListClass *g_render_effects2;
C_RenderTransitionClass *g_render_transition;
C_RLibrary *g_render_library;

int is_mmx(void) {	
	DWORD retval1,retval2;	
	__try { 		
		_asm {
			mov eax, 1		// set up CPUID to return processor version and features
										//	0 = vendor string, 1 = version info, 2 = cache info
			_emit 0x0f	// code bytes = 0fh,  0a2h
			_emit 0xa2
			mov retval1, eax		
			mov retval2, edx
		}	
	} __except(EXCEPTION_EXECUTE_HANDLER) { retval1 = retval2= 0;}
	if (!retval1) return 0;
	return (retval2&0x800000)?1:0;
}


unsigned char g_blendtable[256][256]; //  (i*j)/255.0 lookup table. used in blending
unsigned int const mmx_blend4_revn[2]={0xff00ff,0xff00ff};//{0x1000100,0x1000100}; <<- this is actually more correct, but we're going for consistency vs. the non-mmx ver-jf
int const mmx_blendadj_mask[2] = { 0xff00ff,0xff00ff};
int const mmx_blend4_zero=0;

void Render_Init(HINSTANCE hDllInstance)
{
#ifdef LASER
  laser_connect();
  g_laser_linelist=createLineList();
#endif
	timingInit();
  {
    int i,j;
    for (j=0;j<256;j++)
  	  for (i=0;i<256;i++)
	  	  g_blendtable[i][j] = (unsigned char)((i / 255.0) * (float)j);
  }

  g_render_library=new C_RLibrary();
  g_render_effects=new C_RenderListClass(1);
  g_render_effects2=new C_RenderListClass(1);
  g_render_transition=new C_RenderTransitionClass();

	char INI_FILE[MAX_PATH];
	char *p=INI_FILE;
	GetModuleFileName(hDllInstance,INI_FILE,sizeof(INI_FILE));
	if (p[0]) while (p[1]) p++;
	while (p >= INI_FILE && *p != '\\') p--;
#ifdef LASER
	strcpy(p,"\\vis_avs_laser.dat");
#else
	strcpy(p,"\\vis_avs.dat");
#endif
  extern int g_saved_preset_dirty;
  // clear the undo stack before loading a file.
  C_UndoStack::clear();
  g_render_effects->__LoadPreset(INI_FILE,1);
  // then add the new load to the undo stack but mark it clean if it is supposed to be
  C_UndoStack::saveundo();
  if (!g_saved_preset_dirty)
  {
    C_UndoStack::cleardirty();
  }
}

void Render_Quit(HINSTANCE hDllInstance)
{
  if (g_render_transition) delete g_render_transition;
  g_render_transition=NULL;
	if (g_render_effects)
  {
		char INI_FILE[MAX_PATH];
		char *p=INI_FILE;
		GetModuleFileName(hDllInstance,INI_FILE,sizeof(INI_FILE));
		if (p[0]) while (p[1]) p++;
		while (p >= INI_FILE && *p != '\\') p--;
  #ifdef LASER
	  strcpy(p,"\\vis_avs_laser.dat");
  #else
	  strcpy(p,"\\vis_avs.dat");
  #endif

    g_render_effects->__SavePreset(INI_FILE);
	}

  if (g_render_effects) delete g_render_effects;
  g_render_effects=NULL;
  if (g_render_effects2) delete g_render_effects2;
  g_render_effects2=NULL;

  if (g_render_library) delete g_render_library;
  g_render_library=NULL;

	timingPrint();
#ifdef LASER
  if (g_laser_linelist) delete g_laser_linelist;
  g_laser_linelist=0;
  laser_disconnect();
#endif
}


vis_avs/main.cpp
/*
  LICENSE
  -------
Copyright 2005 Nullsoft, Inc.
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, 
are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer. 

  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution. 

  * Neither the name of Nullsoft nor the names of its contributors may be used to 
    endorse or promote products derived from this software without specific prior written permission. 
 
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR 
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND 
FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

*/
#include <windows.h>
#include <math.h>
#include <process.h>
#include "draw.h"
#include "wnd.h"
#include "r_defs.h"
#include "render.h"
#include "vis.h"
#include "cfgwnd.h"
#include "resource.h"
#include "bpm.h"

#include <stdio.h>

#ifdef REAPLAY_PLUGIN
#include "../../jmde/reaper_plugin.h"
const char *(*get_ini_file)();
int (*vuGetVisData)(char *vdata, int size);
#endif

#include "avs_eelif.h"

extern void GetClientRect_adj(HWND hwnd, RECT *r);

static unsigned char g_logtab[256]; // log function lookup table
HINSTANCE g_hInstance;

char *verstr=
#ifndef LASER
"Cockos Happy AVS"
#else
"AVS/Laser"
#endif
" v2.91"
;

static unsigned int WINAPI RenderThread(LPVOID a);

static void config(struct winampVisModule *this_mod);
static int init(struct winampVisModule *this_mod);
static int render(struct winampVisModule *this_mod);
static void quit(struct winampVisModule *this_mod);

HANDLE g_hThread;
volatile int g_ThreadQuit;

static CRITICAL_SECTION g_cs;

static unsigned char g_visdata[2][2][576]; // stores visualization data [spectrum/waveform][channel][samples]
static int g_visdata_pstat;

static winampVisModule *getModule(int which);
static winampVisHeader hdr = { VIS_HDRVER, verstr, getModule };

extern "C" {
	__declspec( dllexport ) winampVisHeader* winampVisGetHeader()
	{
		return &hdr;
	}
}

/**
 * Plugin entry point? Returns a struct containing
 * function pointers to plugin operations and other meta
 * info
 */
static winampVisModule *getModule(int which)
{
	static winampVisModule mod =
	{
#ifdef LASER
		"Advanced Visualization Studio/Laser",
#else
		"Advanced Visualization Studio",
#endif
		NULL,	// hwndParent
		NULL,	// hDllInstance
		0,		// sRate
		0,		// nCh
		1000/70,		// latencyMS
		1000/70,// delayMS
		2,		// spectrumNch
		2,		// waveformNch
		{ 0, },	// spectrumData
		{ 0, },	// waveformData
		config,
		init,
		render, 
		quit
	};
	if (which==0) return &mod;
	return 0;
}

BOOL CALLBACK aboutProc(HWND hwndDlg, UINT uMsg, WPARAM wParam,LPARAM lParam)
{
  switch (uMsg)
  {
    case WM_INITDIALOG: 
      SetDlgItemText(hwndDlg,IDC_VERSTR,verstr);
      
    return 1;
    case WM_COMMAND:
      switch (LOWORD(wParam))
      {
        case IDOK: case IDCANCEL:
          EndDialog(hwndDlg,0);
        return 0;
      }
    return 0;
  }
  return 0;
}

static void config(struct winampVisModule *this_mod)
{
  if (!g_hwnd || !IsWindow(g_hwnd))
  {
    DialogBox(this_mod->hDllInstance,MAKEINTRESOURCE(IDD_DIALOG2),this_mod->hwndParent,aboutProc);
  }
  else
  {
    SendMessage(g_hwnd,WM_USER+33,0,0);
  }
}

CRITICAL_SECTION g_render_cs;
static int g_is_beat;
char g_path[1024];

int beat_peak1,beat_peak2, beat_cnt,beat_peak1_peak;

void main_setRenderThreadPriority()
{
  int prios[]={
    GetThreadPriority(GetCurrentThread()),
    THREAD_PRIORITY_IDLE,
    THREAD_PRIORITY_LOWEST,
    THREAD_PRIORITY_NORMAL,
    THREAD_PRIORITY_HIGHEST,
  };
	SetThreadPriority(g_hThread,prios[cfg_render_prio]);
}

extern void previous_preset(HWND hwnd);
extern void next_preset(HWND hwnd);
extern void random_preset(HWND hwnd);

#if 0//syntax highlighting
HINSTANCE hRich;
#endif

static int init(struct winampVisModule *this_mod)
{
	DWORD id;
  FILETIME ft;
#if 0//syntax highlighting
  if (!hRich) hRich=LoadLibrary("RICHED32.dll");
#endif
  GetSystemTimeAsFileTime(&ft);
  srand(ft.dwLowDateTime|ft.dwHighDateTime^GetCurrentThreadId());
	g_hInstance=this_mod->hDllInstance;
	GetModuleFileName(g_hInstance,g_path,MAX_PATH);
	char *p=g_path+strlen(g_path);
	while (p > g_path && *p != '\\') p--;
	*p = 0;

#ifdef WA2_EMBED
  if (SendMessage(this_mod->hwndParent,WM_USER,0,0) < 0x2900)
  {
    MessageBox(this_mod->hwndParent,"This version of AVS requires Winamp 2.9+","AVS ERROR",MB_OK|MB_ICONSTOP);
    return 1;
  }
#endif

#ifndef NO_MMX
  extern int is_mmx(void);
  if (!is_mmx())
  {
    MessageBox(this_mod->hwndParent,"NO MMX SUPPORT FOUND - CANNOT RUN AVS - GET THE NON-MMX VERSION.","AVS ERROR",MB_OK|MB_ICONSTOP);
    return 1;
  }
#endif

#ifdef LASER
  strcat(g_path,"\\avs_laser");
#else
  strcat(g_path,"\\avs");
#endif
  CreateDirectory(g_path,NULL);

	InitializeCriticalSection(&g_cs);
	InitializeCriticalSection(&g_render_cs);
	g_ThreadQuit=0;
	g_visdata_pstat=1;

	// Expression evaluator init
	AVS_EEL_IF_init();

	// UI init
	if (Wnd_Init(this_mod)) return 1;

	{
		// init the log lookup table. base 60??
		int x;
		for (x = 0; x < 256; x ++)
		{
			double a=log(x*60.0/255.0 + 1.0)/log(60.0);
			int t=(int)(a*255.0);
			if (t<0)t=0;
			if (t>255)t=255;
			g_logtab[x]=(unsigned char )t;
		}
	}

	// init bpm refiner
	initBpm();

	// init the rendering
	Render_Init(g_hInstance);

	//create the editor window
	CfgWnd_Create(this_mod);

	g_hThread=(HANDLE)_beginthreadex(NULL,0,RenderThread,0,0,(unsigned int *)&id);
	main_setRenderThreadPriority();
	SetForegroundWindow(g_hwnd);
	SetFocus(g_hwnd);

	return 0;
}

static int render(struct winampVisModule *this_mod)
{
	int x,avs_beat=0,b;
	if (g_ThreadQuit) return 1;
  EnterCriticalSection(&g_cs);
	if (g_ThreadQuit)
	{
		LeaveCriticalSection(&g_cs);
		return 1;
	}

	// put spectrum data in log scale
	if (g_visdata_pstat)
		for (x = 0; x<  576*2; x ++)
			g_visdata[0][0][x]=g_logtab[(unsigned char)this_mod->spectrumData[0][x]];
	else 
	{
		for (x = 0; x < 576*2; x ++)
		{ 
			int t=g_logtab[(unsigned char)this_mod->spectrumData[0][x]];
			if (g_visdata[0][0][x] < t)
				g_visdata[0][0][x] = t;
		}
	}
	memcpy(&g_visdata[1][0][0],this_mod->waveformData,576*2);
	{
    int lt[2]={0,0};
    int x;
    int ch;
    for (ch = 0; ch < 2; ch ++)
    {
      unsigned char *f=(unsigned char*)&this_mod->waveformData[ch][0];
      for (x = 0; x < 576; x ++)
      {
        int r= *f++^128;
        r-=128;
        if (r<0)r=-r;
        lt[ch]+=r;
      }
    }
    lt[0]=max(lt[0],lt[1]);

    beat_peak1=(beat_peak1*125+beat_peak2*3)/128;

    beat_cnt++;

    if (lt[0] >= (beat_peak1*34)/32 && lt[0] > (576*16)) 
    {
      if (beat_cnt>0)
      {
        beat_cnt=0;
        avs_beat=1;
      }
      beat_peak1=(lt[0]+beat_peak1_peak)/2;
      beat_peak1_peak=lt[0];
    }
    else if (lt[0] > beat_peak2)
    {
      beat_peak2=lt[0];
    } 
    else beat_peak2=(beat_peak2*14)/16;

	}
	b=refineBeat(avs_beat);
	if (b) g_is_beat=1;
	g_visdata_pstat=0;
	LeaveCriticalSection(&g_cs);
  return 0;
}

static void quit(struct winampVisModule *this_mod)
{
#define DS(x) 
  //MessageBox(this_mod->hwndParent,x,"AVS Debug",MB_OK)
	if (g_hThread)
	{
    DS("Waitin for thread to quit\n");
		g_ThreadQuit=1;
		if (WaitForSingleObject(g_hThread,10000) != WAIT_OBJECT_0)
		{
      DS("Terminated thread (BAD!)\n");
			//MessageBox(NULL,"error waiting for thread to quit","a",MB_TASKMODAL);
      TerminateThread(g_hThread,0);
		}
    DS("Thread done... calling ddraw_quit\n");
		DDraw_Quit();

    DS("Calling cfgwnd_destroy\n");
		CfgWnd_Destroy();
    DS("Calling render_quit\n");
		Render_Quit(this_mod->hDllInstance);

    DS("Calling wnd_quit\n");
		Wnd_Quit();

    DS("closing thread handle\n");
		CloseHandle(g_hThread);
		g_hThread=NULL;

    DS("calling eel quit\n");
    AVS_EEL_IF_quit();

    DS("cleaning up critsections\n");
		DeleteCriticalSection(&g_cs);
		DeleteCriticalSection(&g_render_cs);    

    DS("smp_cleanupthreads\n");
    C_RenderListClass::smp_cleanupthreads();
	}
#undef DS
#if 0//syntax highlighting
  if (hRich) FreeLibrary(hRich);
  hRich=0;
#endif
}

#define FPS_NF 64

static unsigned int WINAPI RenderThread(LPVOID a)
{
  int framedata[FPS_NF]={0,};
	int framedata_pos=0;
  int s=0;
	char vis_data[2][2][576];
  FILETIME ft;
  GetSystemTimeAsFileTime(&ft);
  srand(ft.dwLowDateTime|ft.dwHighDateTime^GetCurrentThreadId());
	while (!g_ThreadQuit)
	{
		int w,h,*fb=NULL, *fb2=NULL,beat=0;

#ifdef REAPLAY_PLUGIN
    if(!IsWindowVisible(g_hwnd)) 
    {
      Sleep(1);
      continue;
    }

    char visdata[576*2*2];
    int ret = vuGetVisData(visdata, sizeof(visdata));
		if (!ret) 
    {
      memset(&vis_data[0][0][0],0,576*2*2);
      beat=0;
    }
    else
    {
      int x;
      unsigned char *v=(unsigned char *)visdata;
		  for (x = 0; x < 576*2; x ++)
			  vis_data[0][0][x]=g_logtab[*v++];
		  for (x = 0; x < 576*2; x ++)
        ((unsigned char *)vis_data[1][0])[x]=*v++;

      v=(unsigned char *)visdata+1152;
	    {
        int lt[2]={0,0};
        int ch;
        for (ch = 0; ch < 2; ch ++)
        {
          for (x = 0; x < 576; x ++)
          {
            int r=*v++^128;
            r-=128;
            if (r<0)r=-r;
            lt[ch]+=r;
          }
        }
        lt[0]=max(lt[0],lt[1]);

        beat_peak1=(beat_peak1*125+beat_peak2*3)/128;
        beat_cnt++;

        if (lt[0] >= (beat_peak1*34)/32 && lt[0] > (576*16)) 
        {
          if (beat_cnt>0)
          {
            beat_cnt=0;
            beat=1;
          }
          beat_peak1=(lt[0]+beat_peak1_peak)/2;
          beat_peak1_peak=lt[0];
        }
        else if (lt[0] > beat_peak2)
        {
          beat_peak2=lt[0];
        } 
        else beat_peak2=(beat_peak2*14)/16;

	    }
//     EnterCriticalSection(&g_title_cs);
	    beat=refineBeat(beat);
//      LeaveCriticalSection(&g_title_cs);
    }
#else
		EnterCriticalSection(&g_cs);
		memcpy(&vis_data[0][0][0],&g_visdata[0][0][0],576*2*2);
		g_visdata_pstat=1;
		beat=g_is_beat;
		g_is_beat=0;
		LeaveCriticalSection(&g_cs);
#endif

    if (!g_ThreadQuit)
    {
		  if (IsWindow(g_hwnd)&&!g_in_destroy) DDraw_Enter(&w,&h,&fb,&fb2);
      else break;
		  if (fb&&fb2)
		  {
        extern int g_dlg_w, g_dlg_h, g_dlg_fps;
#ifdef LASER
        g_laser_linelist->ClearLineList();
#endif

	      EnterCriticalSection(&g_render_cs);
				int t=g_render_transition->render(vis_data,beat,s?fb2:fb,s?fb:fb2,w,h);
	      LeaveCriticalSection(&g_render_cs);
        if (t&1) s^=1;

#ifdef LASER
        s=0;
        memset(fb,0,w*h*sizeof(int));
        LineDrawList(g_laser_linelist,fb,w,h);
#endif
			  if (IsWindow(g_hwnd)) DDraw_Exit(s);

        int lastt=framedata[framedata_pos];
        int thist=GetTickCount();
        framedata[framedata_pos]=thist;
        g_dlg_w=w;
        g_dlg_h=h;
        if (lastt)
        {
          g_dlg_fps=MulDiv(sizeof(framedata)/sizeof(framedata[0]),10000,thist-lastt);
        }
        framedata_pos++;
        if (framedata_pos >= sizeof(framedata)/sizeof(framedata[0])) framedata_pos=0;

		  }
      int fs=DDraw_IsFullScreen();
      int sv=(fs?(cfg_speed>>8):cfg_speed)&0xff;
	    Sleep(min(max(sv,1),100));
    }
	}
  _endthreadex(0);
	return 0;
}

#ifdef REAPLAY_PLUGIN
static winampVisModule dummyMod;
extern "C"
{
  
  REAPER_PLUGIN_DLL_EXPORT int REAPER_PLUGIN_ENTRYPOINT(REAPER_PLUGIN_HINSTANCE hInstance, reaper_plugin_info_t *rec)
  {
    g_hInstance=hInstance;
    if (rec)
    {
      if (rec->caller_version != REAPER_PLUGIN_VERSION || !rec->GetFunc)
        return 0;

      *((void **)&get_ini_file) = rec->GetFunc("get_ini_file");
      *((void **)&vuGetVisData) = rec->GetFunc("vuGetVisData");
      if (!get_ini_file || !vuGetVisData)
        return 0;

      dummyMod.hwndParent=rec->hwnd_main;
      dummyMod.hDllInstance=g_hInstance;
      init(&dummyMod);

      return 1;
    }
    else
    {
      quit(&dummyMod);
      return 0;
    }
  }
  
};
#endif
